lms/djangoapps/grades/models.py

```diff
@@ -0,0 +1,322 @@
+
+    def __unicode__(self):
+        """
+        Returns a string representation of this model.
+        """
+        return u"{} user: {}, course version: {}, subsection {} ({}). {}/{} graded, {}/{} all".format(
+            type(self).__name__,
+            self.user_id,
+            self.course_version,
+            self.usage_key,
+            self.visible_blocks.hashed,
+            self.earned_graded,
+            self.possible_graded,
+            self.earned_all,
+            self.possible_all,
+        )
+
+    @classmethod
+    def save_grade(cls, **kwargs):
+        """
+        Wrapper for create_grade or update_grade, depending on which applies.
+        Takes the same arguments as both of those methods.
+        """
+        user_id = kwargs.pop('user_id')
+        usage_key = kwargs.pop('usage_key')
+        try:
+            grade, is_created = cls.objects.get_or_create(user_id=user_id, usage_key=usage_key, defaults=kwargs)
+        except IntegrityError:
+            is_created = False
+        if not is_created:
+            grade.update(**kwargs)
+
+    @classmethod
+    def read_grade(cls, user_id, usage_key):
+        """
+        Reads a grade from database
+
+        Arguments:
+            user_id: The user associated with the desired grade
+            usage_key: The location of the subsection associated with the desired grade
+
+        Raises PersistentSubsectionGrade.DoesNotExist if applicable
+        """
+        return cls.objects.get(
+            user_id=user_id,
+            course_id=usage_key.course_key,  # course_id is included to take advantage of db indexes
+            usage_key=usage_key,
+        )
+
+    @classmethod
+    def update_grade(
+            cls,
+            user_id,
+            usage_key,
+            course_version,
+            subtree_edited_date,
+            earned_all,
+            possible_all,
+            earned_graded,
+            possible_graded,
+            visible_blocks,
+    ):
+        """
+        Updates a previously existing grade.
+
+        Requires all the arguments listed in docstring for create_grade
+        """
+        grade = cls.read_grade(
+            user_id=user_id,
+            usage_key=usage_key,
+        )
+
+        grade.update(
+            course_version=course_version,
+            subtree_edited_date=subtree_edited_date,
+            earned_all=earned_all,
+            possible_all=possible_all,
+            earned_graded=earned_graded,
+            possible_graded=possible_graded,
+            visible_blocks=visible_blocks,
+        )
+
+    def update(
+            self,
+            course_version,
+            subtree_edited_date,
+            earned_all,
+            possible_all,
+            earned_graded,
+            possible_graded,
+            visible_blocks,
+    ):
+        """
+        Modify an existing PersistentSubsectionGrade object, saving the new
+        version.
+        """
```

Comment:
Can you add a comment on the difference between `update` and `update_grade` and when to use which one?